internal/resource/registry.go

```diff
@@ -0,0 +1,58 @@
+package resource
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/proto-public/pbresource"
+)
+
+type Registry interface {
+	// Register the given resource type and its hooks.
+	Register(reg Registration)
+
+	// Resolve the given resource type and its hooks.
+	Resolve(typ *pbresource.Type) (reg Registration, ok bool)
+}
+
+type Registration struct {
+	// Type is the GVK of the resource type.
+	Type *pbresource.Type
+
+	// In the future, we'll add hooks, the controller etc. here.
+}
+
+// Resource type registry
+type TypeRegistry struct {
+	registrations map[*pbresource.Type]Registration
+}
+
+func NewRegistry() Registry {
+	return &TypeRegistry{
+		registrations: make(map[*pbresource.Type]Registration),
+	}
+}
+
+func (r *TypeRegistry) Register(registration Registration) {
+	typ := registration.Type
+	if typ.Group == "" || typ.GroupVersion == "" || typ.Kind == "" {
+		panic("type field(s) cannot be empty")
+	}
+
+	if _, ok := r.registrations[registration.Type]; ok {
+		panic(fmt.Sprintf("resource type %s already registered", ToGVK(registration.Type)))
+	}
+
+	r.registrations[registration.Type] = registration
```

Comment:
I think it might be unsafe to use `Type` as a map key, because it has a bunch of internal protobuf fields which would presumably change the hash value. Could we use the `ToGVK` string instead?

Reply:
👍 Fixed in https://github.com/hashicorp/consul/pull/16622/commits/c5b3faa0c528d2154e42cd3f7ce88d88aa95535a